src/System.Private.CoreLib/shared/System/Index.cs

```diff
@@ -20,17 +36,87 @@ public Index(int value, bool fromEnd)
             _value = fromEnd ? ~value : value;
         }
 
+        // The following private constructors mainly created for perf reason to avoid the checks
+        private Index(int value)
+        {
+            _value = value;
+        }
+
+        /// <summary>Create an Index pointing at first element.</summary>
+        public static Index Start => new Index(0);
+
+        /// <summary>Create an Index pointing at beyond last element.</summary>
+        public static Index End => new Index(~0);
+
+        /// <summary>Create an Index from the start at the position indicated by the value.</summary>
+        /// <param name="value">The index value from the start.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Index FromStart(int value)
+        {
+            if (value < 0)
+            {
+                ThrowHelper.ThrowValueArgumentOutOfRange_NeedNonNegNumException();
+            }
+
+            return new Index(value);
+        }
+
+        /// <summary>Create an Index from the end at the position indicated by the value.</summary>
+        /// <param name="value">The index value from the end.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Index FromEnd(int value)
+        {
+            if (value < 0)
+            {
+                ThrowHelper.ThrowValueArgumentOutOfRange_NeedNonNegNumException();
+            }
+
+            return new Index(~value);
+        }
+
+        /// <summary>Returns the index value.</summary>
         public int Value => _value < 0 ? ~_value : _value;
-        public bool FromEnd => _value < 0;
+
+        /// <summary>Indicates whether the index is from the start or the end.</summary>
+        public bool IsFromEnd => _value < 0;
+
+        /// <summary>Calculate the offset from the start using the giving collection length.</summary>
+        /// <param name="length">The length of the collection that the Index will be used with. length has to be a positive value</param>
+        public int GetOffset(int length)
+        {
+            if (length < 0)
+            {
+                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
+            }
+
+            int offset = IsFromEnd ? length - (~_value) : _value;
+            if ((uint)offset > (uint)length)
+            {
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
```

Comment:
If this is used in an indexer, is it going to be strange that it results in an Argument out of range rather than IndexOutOfRange?

I'm wondering if maybe this shouldn't do any validation, just the math, and leave it up to the caller whether and how to validate. That could actually be cheaper, as they generally won't need to do the first check about the length being negative.

Reply:
actually I prefer Argument exception with ExceptionResource.Argument_InvalidOffLen as it is telling the passed length is not scoping the index. Also, I prefer to keep the validation because we can end up with negative offset which would be bad for consumers.

Reply:
> actually I prefer Argument exception with ExceptionResource.Argument_InvalidOffLen as it is telling the passed length is not scoping the index

Sure, but where is this API going to be used?  Generally we throw IndexOutOfRange exception for indexers, so if this is going to be used in an indexer, and someone wants to conform to throwing IndexOutOfRange, they won't be able to use this API, right?

> I prefer to keep the validation because we can end up with negative offset which would be bad for consumers

In the 99% case, wouldn't the index then just be passed into the indexer of an array which is itself going to validate it and throw an IndexOutOfRangeException?

Reply:
> are you suggesting to remove all validations? including the length in the beginning of the method?

Would we use it if we didn't remove all the validation?  Would we use it if we did?

But yes, that's what I'm asking about, removing all the validation, just doing the math, and leaving it up to the caller to determine whether the resulting offset makes sense.  Certainly there are cases where the result could be "valid" if given non-sensical input, e.g. provided with a negative length and a positive enough offset to make the result non-negative, but based on how this API will actually be used, does that matter?  It seems like the 99% use case will be in writing wrapper indexers like:
```C#
class MyCollection<T>
{
    public T this[int index] { … } // existing
    public T this[Index index] => this[index.GetOffset(Count)]; // new
}
```
in which case the caller is going to be passing in a valid length by construction, and the delegated indexer is going to be doing its own validation that ensures the resulting actual index is in range.  If nothing else, by not validating the input, we get out of the business of determining what the right exception type is to throw when this helper is used in the scope of these larger methods.

Reply:
thinking more, I think your suggestion is good. if later anyone needed a method to validate, then we can provide another one at that time. I'll apply your suggestion for both Index.GetOffset and Range.GetOffsetLenth too. Thank you.

Reply:
done that in the latest iteration